### PR TITLE
ARROW-15594: [C++][FlightRPC] Add Deserialize(const Buffer&) to various Flight types

### PR DESCRIPTION
**Pull request by lidavidm**

### CODE REVIEW

File: cpp/src/arrow/flight/types.cc [hunk 1]
```diff
@@ -234,6 +260,23 @@ Status FlightInfo::Deserialize(const std::string& serialized,
   return Status::OK();
 }
 
+Status FlightInfo::Deserialize(const Buffer& serialized,
+                               std::unique_ptr<FlightInfo>* out) {
```
Comment (cyb70289) [inline cpp/src/arrow/flight/types.cc:264]: Nit: Is it possible to combine these XXX:Deserialize functions into one, if change this unique_ptr to raw pointer as other functions do.

**Review by cyb70289 — APPROVED** [review]
+1

**Comment by pitrou** [comment]
Perhaps it would be simpler to just take a `string_view` rather than have two overloads?

**Comment by lidavidm** [comment]
Good point - consolidated on string_view, also changed to use Result while I was at it.

File: cpp/src/arrow/flight/types.h [hunk 2]
```diff
@@ -165,8 +166,15 @@ struct ARROW_FLIGHT_EXPORT BasicAuth {
...
   std::string password;
 
-  static Status Deserialize(const std::string& serialized, BasicAuth* out);
+  /// \brief Deserialize this message from its wire-format representation.
+  static arrow::Result<BasicAuth> Deserialize(const arrow::util::string_view& serialized);
+  /// \brief Serialize this message to its wire-format representation.
+  arrow::Result<std::string> SerializeToString() const;
 
+  ARROW_DEPRECATED("Deprecated in 8.0.0. Use Result-returning overload instead.")
+  static Status Deserialize(const arrow::util::string_view& serialized, BasicAuth* out);
```
Comment (pitrou) [inline cpp/src/arrow/flight/types.h:175]: Hmm, not sure we should change the signature (from `string` to `string_view`) while deprecating it? Same remark for below as well.

Comment (lidavidm) [reply cpp/src/arrow/flight/types.h]: Good point - undid that change (I refactored the signature, then added the deprecation, should've done it the other way around.)

File: cpp/src/arrow/flight/types.h [hunk 3]
```diff
@@ -165,8 +166,15 @@ struct ARROW_FLIGHT_EXPORT BasicAuth {
   std::string username;
   std::string password;
 
+  /// \brief Deserialize this message from its wire-format representation.
+  static arrow::Result<BasicAuth> Deserialize(const arrow::util::string_view& serialized);
```
Comment (cyb70289) [inline cpp/src/arrow/flight/types.h:170]: Prefer passing by value (in registers) for trivial objects like `string_view`?
If passing by reference, the callee has to deference the pointer.

Comment (lidavidm) [reply cpp/src/arrow/flight/types.h]: Thanks for the catch - fixed.

**Review by pitrou — APPROVED** [review]

**Pull request closed without merging**